### DISCUSSION TITLE
Implement PREFERRED_PHONE_ACCOUNT_ID

### DISCUSSION THREAD

**Discussion by NLLAPPS** [vestrel00/contacts-android#247] [Ideas]
As per https://developer.android.com/reference/android/provider/ContactsContract.DataColumns#PREFERRED_PHONE_ACCOUNT_ID

**Comment by vestrel00** [comment]
@NLLAPPS based on my preliminary investigations, the `PREFERRED_PHONE_ACCOUNT_ID` is used by dialer apps in conjunction with `PhoneAccountHandle` and `PhoneAccount`. All of these have been recently introduced in Android API 28. It seems like an advanced feature that is very specific for dialer apps (mainly dealing with `TelecomManager`), which I'm currently not prioritizing. 

![Screen Shot 2022-07-06 at 10 51 44 AM](https://user-images.githubusercontent.com/4332753/177641036-84214a93-d612-4f0f-a552-82fa7520d0fb.png)
![Screen Shot 2022-07-06 at 10 51 52 AM](https://user-images.githubusercontent.com/4332753/177641047-fbaa04cf-b67e-4599-a036-cac2dfb501b4.png)
![Screen Shot 2022-07-06 at 10 51 58 AM](https://user-images.githubusercontent.com/4332753/177641060-079e9a54-14e4-4e43-bb6f-29bb4978a952.png)

I can certainly add in that field as part of the `Phone` entity but I'm not sure what exactly it is used for and how. It's going to take me days to investigate and document. It's probably related to a bunch of other things as I pointed out above. 

--------------

I'll wait to add this until I get more time to investigate and until I complete more higher priority issues in the **[Project Roadmap](https://github.com/vestrel00/contacts-android/wiki/Project-Roadmap#additional-features-for-v1)**.

For now, I will convert this to a discussion so that other people can participate if they are also interested 😁 

Thanks!

**Comment by NLLAPPS** [comment]
Yes it is specific to dialers. It is essentially another data where a dialer can get the phone account handle to pass to system when dialing. Only applies to dual sim/multi phone account devices. I can try to answer any questions you may have. 

I am considering to move to your library for my planned contact editing feature. It looks to be a very complicated path if I go towards vanilla contentresolver route.

**Reply by vestrel00** [reply]
> I am considering to move to your library for my planned contact editing feature.

AWESOME! I'd love to have you as a user of this library. 

> Yes it is specific to dialers. Only applies to dual sim/multi phone account devices.

Unfortunately, dialer support is not the main focus of this library. This library is focused on contacts, which is related to dialers.

I think you should still be able to use this library even without the addition of `PREFERRED_PHONE_ACCOUNT_ID`.

>  I can try to answer any questions you may have.

I can still add this field/property to the library without fully supporting everything related to it. Just answer me a few questions;

1. Can you set this field while using the default Contacts app or dialer/phone app? If so, how do you do it? (A screen recording would be great, but optional, so I can see).
2. Is adding the `PREFERRED_PHONE_ACCOUNT_ID` field/property to this library's [`PhoneEntity`](https://github.com/vestrel00/contacts-android/blob/main/core/src/main/java/contacts/core/entities/Phone.kt) all you need in order for you to use this library in your app?
3. Please log the Data table so I can see some non-null values for this column.

```kotlin
val cursor = contentResolver.query(
    ContactsContract.Data.CONTENT_URI,
    arrayOf(
        ContactsContract.Data._ID,
        ContactsContract.Data.RAW_CONTACT_ID,
        ContactsContract.Data.CONTACT_ID,
        ContactsContract.Data.MIMETYPE,
        ContactsContract.Data.PREFERRED_PHONE_ACCOUNT_ID,
        ContactsContract.Data.IS_PRIMARY,
        ContactsContract.Data.IS_SUPER_PRIMARY
    ),
    null,
    null,
    null
)

while (cursor.moveToNext()) {
    // Use getString instead of getLong, getInt, etc so that the value could be null.
    val id = cursor.getString(0)
    val rawContactId = cursor.getString(1)
    val contactId = cursor.getString(2)
    val mimeType = cursor.getString(3)
    val preferredPhoneAccountId = cursor.getString(4)
    val isPrimary = cursor.getString(5)
    val isSuperPrimary = cursor.getString(6)

    log(
        """
            Data id: $id, rawContactId: $rawContactId, contactId : $contactId,
             mimeType: $mimeType, preferredPhoneAccountId: $preferredPhoneAccountId,
             isPrimary: $isPrimary, isSuperPrimary: $isSuperPrimary
        """.trimIndent().replace("\n", "")
    )
}

cursor.close()
```

**Comment by NLLAPPS** [comment]
1- It requires Contact permission only. I don't have dual SIM device to test it at the moment.
2- Most probably. Should be updatable though.
3- I am not using your library nor implemented PREFERRED_PHONE_ACCOUNT_ID  yet. 
I think ContactsContract.CommonDataKinds.Phone.CONTENT_URI should be used for this otherwise you would get all available mime types not just phone. A quick test however returns null for PREFERRED_PHONE_ACCOUNT_ID

**Reply by vestrel00** [reply]
> 1. Can you set this field while using the default Contacts app or dialer/phone app? If so, how do you do it? (A screen recording would be great, but optional, so I can see).
> > 1- It requires Contact permission only. I don't have dual SIM device to test it at the moment.

I don't have one either. I will need your help to do the investigation. I will probably guide you in contributing to this library by adding this field/property yourself 😁 

We will have to wait until you are able to test this yourself. Let me know when you are ready.

> 2. Is adding the `PREFERRED_PHONE_ACCOUNT_ID` field/property to this library's [PhoneEntity](https://github.com/vestrel00/contacts-android/blob/main/core/src/main/java/contacts/core/entities/Phone.kt) all you need in order for you to use this library in your app?
> > 2- Most probably. Should be updatable though.

Yes. Full CRUD (create, read, update, delete) operations should be supported.

> 3. Please log the Data table so I can see some non-null values for this column.
> > 3. I am not using your library nor implemented PREFERRED_PHONE_ACCOUNT_ID yet. I think ContactsContract.CommonDataKinds.Phone.CONTENT_URI should be used for this...

Part of the investigation is not ruling out any possibility until you've seen it for yourself. I intentionally asked you to **log all rows in the data table by using `ContactsContract.Data.CONTENT_URI`** because of the possibility that `PREFERRED_PHONE_ACCOUNT_ID`;

- is only available (non-null) for just phone mime types
- is available (non-null) in all data rows and not just phone mime types
- is not using any common data kinds mime types such as phone mime type and might have a special value in the mime type column that may be unique per account

I'm hoping that it "is only available (non-null) for just phone mime types". However, we must not rule out the other possibilities until we can prove it **by logging all data rows**. 

I'm suspicious because of the location of `PREFERRED_PHONE_ACCOUNT_ID`...

```java
ContactsContract {
    DataColumns {
        PREFERRED_PHONE_ACCOUNT_ID
    }
    CommonDataKinds {
        Phone {
            CONTENT_ITEM_TYPE, CONTENT_URI, NUMBER, NORMALIZED_NUMBER
        }
    }
}
```

If this field/column is supposed to only be available for Data table rows of phone mime types, then why is it not placed in `CommonDataKinds.Phone`?

We need to answer the following questions;

1. How many different `PREFERRED_PHONE_ACCOUNT_ID ` can one **RawContact** have? Only one or possibly more? 
2. Can every phone mime type rows in Data table have a different value for `PREFERRED_PHONE_ACCOUNT_ID`?
3. Is the `PREFERRED_PHONE_ACCOUNT_ID` associated with a **RawContact** (Account-dependent) or the parent **Contact** (not dependent on Account)?

Let's continue this discussion when you are able to test. Thanks! 🍻
